plugins/chain_plugin/get_info_db.cpp

```diff
@@ -0,0 +1,89 @@
+#include <eosio/chain_plugin/get_info_db.hpp>
+#include <eosio/chain/resource_limits.hpp>
+#include <eosio/chain/application.hpp>
+
+#include <shared_mutex>
+
+using namespace eosio;
+using namespace appbase;
+
+namespace eosio::chain_apis {
+   /**
+    * Implementation details of the get_info results
+    */
+   struct get_info_db_impl {
+      get_info_db_impl(const chain::controller& controller, bool get_info_enabled)
+         : controller(controller)
+         , get_info_enabled(get_info_enabled) {}
+
+      // A handle to the controller.
+      const chain::controller& controller;
+
+      // Indication whether get_info RPC is enabled.
+      bool get_info_enabled = false;
+
+      // Cache to store the current get_info results in respect to latest accepted block.
+      get_info_db::get_info_results  cached_results;
+      mutable std::shared_mutex      rw_mutex;
+
+      // Called on accepted_block signal.
+      void on_accepted_block() {
+         try {
+            if (!get_info_enabled) {
+               return;
+            }
+
+            std::unique_lock write_lock(rw_mutex);
+
+            const auto& rm       = controller.get_resource_limits_manager();
+            auto        head_id  = controller.head().id();
+            auto        lib_id   = controller.last_irreversible_block_id();
+            auto        fhead_id = controller.fork_db_head().id();
+
+            // cache get_info results
+            cached_results = get_info_db::get_info_results {
+               itoh(static_cast<uint32_t>(app().version())),
```

Comment:
I doubt it makes much difference, but some of these values never change and could be only set on the first call.